qwertytrewq | 2023-07-28 14:18:57 UTC | #1

Running https://github.com/vporton/NacDB/blob/version3/stress-test/motoko/main.mo by `make stress-test` in `stress-test/` directory I get weird error message (below). Please explain what it means and how to get rid of this error.

```
dfx canister call stresser main '()'
Error: Failed update call.
Caused by: Failed update call.
  The replica returned a replica error: Replica Error: reject code CanisterReject, reject message IC0504: Canister be2us-64aaa-aaaaa-qaabq-cai violated contract: "ic0_msg_cycles_available128" cannot be executed in replicated query mode, error code None
```

-------------------------

pramitgaha | 2023-07-28 20:42:03 UTC | #2

are you trying to perform inter canister calls from the method marked as query?

-------------------------

qwertytrewq | 2023-07-29 01:12:00 UTC | #3

Probably. But why then the error is not caught at compilation stage?

-------------------------

qwertytrewq | 2023-07-29 05:32:20 UTC | #4

Apparently, the error was caused by `ExperimentalCycles.accept()` in a `query` function.

-------------------------

claudio | 2023-07-31 15:10:06 UTC | #5

We would need a stronger type system or dedicated syntax for cycles management. 

The current library based approach leads to dynamic errors but was adopted at the time because the design of cycles on the platform was in flux so we didn't want to prematurely extend the language.

It's probably time to do something better.

-------------------------

